_posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc

```diff
@@ -0,0 +1,2334 @@
+the deployment platforms with varying permissions for handling specific workflows. By separating out the concerns for
+each pipeline, you can reduce the blast radius of the damage that can be done with each set of credentials. At a minimum,
+you should have two versions of the infrastructure deployment system: one for deploying the application code, and which
+only has the minimal permissions necessary for deploying that application; and one for deploying infrastructure code,
+which has full access to the environments.
+
+
+[[use_approval_flows]]
+=== Use approval flows
+
+It is important that human review is baked into each deployment. As covered in <<cicd_for_infrastructure_code>>, it is
+difficult to build an automated test suite that builds enough confidence in your infrastructure code to do the right
+thing. This is important, as failed infrastructure deployments could be catastrophic to your business, and there is no
+concept of rollback with infrastructure deployment tools. This means that you will almost always want to have some form
+of approval workflow for your infrastructure CI/CD pipeline so that you can review what is about to be deployed. Most
+generic CI/CD platforms support approval workflows. For example, CircleCI supports
+https://circleci.com/docs/2.0/workflows/#holding-a-workflow-for-a-manual-approval[approval steps in its workflow
+engine], in addition to https://circleci.com/docs/2.0/contexts/#restricting-a-context[restricted contexts] to limit who
+can approve the workflow.
+
+
+[[lock_down_vcs_systems]]
+=== Lock down VCS systems
+
+It is a good practice to define and store the deployment pipeline as code in the same repo that it is used. For example,
+you should define the CI/CD deployment pipeline for your infrastructure code in the `modules` and `live` repositories.
+However, this means that anyone with access to those repositories are free to modify the pipeline, __even on feature
+branches__. This can be exploited to skip any approval process you have defined in the pipeline by creating a new branch
+that overwrites the pipeline configuration.
+
+This is not a concern if only admin users had access to the infrastructure code. Typically, however, many operations
+teams want contributions to the infrastructure code from developers as well, and having any developer have the ability to
+deploy arbitrary infrastructure to production without any review can be undesirable. To mitigate these concerns, you
+should lock down your VCS systems:
+
+Only deploy from protected branches::
+  In most git hosting platforms, there is a concept of protected branches (see
+  https://help.github.com/en/github/administering-a-repository/about-protected-branches[GitHub docs] for example).
+  Protected branches allow you to implement policies for controlling what code can be merged in. For most platforms, you
+  can protect a branch such that: (a) it can never be force pushed, (b) it can never be merged to or commit to from the
```

Comment:
This list might be more readable on separate lines.

Reply:
Agreed, but due to the limitations of asciidoc definition lists it's a bit hard to massage this. I'm going to keep it as is for now and we can revisit later.